<center><b><font size=5>Выпускной проект: SQL</font></b></center>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исследование таблиц</a></span><ul class="toc-item"><li><span><a href="#books" data-toc-modified-id="books-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>books</a></span></li><li><span><a href="#authors" data-toc-modified-id="authors-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>authors</a></span></li><li><span><a href="#publishers" data-toc-modified-id="publishers-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>publishers</a></span></li><li><span><a href="#ratings" data-toc-modified-id="ratings-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>ratings</a></span></li><li><span><a href="#reviews" data-toc-modified-id="reviews-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>reviews</a></span></li></ul></li><li><span><a href="#Выполнение-заданий" data-toc-modified-id="Выполнение-заданий-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Выполнение заданий</a></span><ul class="toc-item"><li><span><a href="#Количество-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-вышедших-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Количество книг, вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-для-каждой-книги-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка для каждой книги</a></span></li><li><span><a href="#Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг.-Учитывать-только-книги-с-50-и-более-оценками." data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг.-Учитывать-только-книги-с-50-и-более-оценками.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг. Учитывать только книги с 50 и более оценками.</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

<center><b><font size="4">Практическая часть</font></b></center>

Подключимся к бд.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'student', # имя пользователя
'pwd': 'student', # пароль
'host': 'student',
'port': 6653, # порт подключения
'db': 'sql-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)
con=engine.connect()

### Исследование таблиц

Для ознакомления с таблицами выведем по 5 строк каждой из них.

#### books

In [2]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''


pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

#### authors

In [3]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM authors LIMIT 5'''


pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

#### publishers

In [4]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM publishers LIMIT 5'''


pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

#### ratings

In [5]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM ratings LIMIT 5'''


pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

#### reviews

In [6]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM reviews LIMIT 5'''


pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Выполнение заданий

#### Количество книг, вышедших после 1 января 2000 года

In [7]:
query = '''SELECT COUNT(*) 
FROM books 
WHERE publication_date > TO_DATE ('2000-01-01', 'YYYY-MM-DD')
'''

pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

#### Количество обзоров и средняя оценка для каждой книги

In [8]:
query = '''SELECT b.book_id, b.title, COUNT(DISTINCT(rev.review_id)) AS review_count, ROUND(AVG(r.rating),2) AS avg_rating
FROM books AS b
LEFT JOIN reviews AS rev ON rev.book_id = b.book_id
LEFT JOIN ratings AS r ON r.book_id = b.book_id
GROUP BY b.book_id
ORDER BY avg_rating DESC, review_count DESC
'''
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  review_count  \
0         17                      A Dirty Job (Grim Reaper  #1)             4   
1        553            School's Out—Forever (Maximum Ride  #2)             3   
2        444       Moneyball: The Art of Winning an Unfair Game             3   
3        642                 The Big Bad Wolf  (Alex Cross  #9)             2   
4        967               Welcome to Temptation (Dempseys  #1)             2   
..       ...                                                ...           ...   
995      915  The World Is Flat: A Brief History of the Twen...             3   
996      202                                      Drowning Ruth             3   
997      316                  His Excellency: George Washington             2   
998      371                                              Junky             2   
999      303                               Harvesting the Heart             2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 4 columns]

#### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query = '''SELECT p.publisher, COUNT(b.book_id) AS book_count
FROM books AS b
LEFT JOIN publishers AS p ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY book_count DESC
LIMIT 1
'''
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  book_count
0  Penguin Books          42

#### Автор с самой высокой средней оценкой книг. Учитывать только книги с 50 и более оценками.

In [10]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''
WITH rating_count AS (SELECT book_id, COUNT(rating) AS rat_count
FROM ratings
GROUP BY book_id
HAVING COUNT(rating) >= 50)

SELECT a.author, ROUND(AVG(r.rating), 2) AS avg_rating
FROM authors AS a
JOIN books AS b ON b.author_id = a.author_id
JOIN rating_count AS rc ON rc.book_id = b.book_id
JOIN ratings AS r ON r.book_id = rc.book_id
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
'''
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

#### Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [11]:
query = '''WITH u_c AS (SELECT username, COUNT(*)
FROM reviews
WHERE username IN (SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(*) > 48)
GROUP BY username)

SELECT AVG(count)
FROM u_c
'''
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

### Вывод

В ходе выполнения заданий было:

- выяснено, что после 1 января 2000 года вышло 819 книг;
- посчитано количество обзоров и средняя оценка для каждой книги;
- выяснено, что "Penguin Books" - издательство, которое выпустило наибольшее число книг толще 50 страниц;
- выяснено, что J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками;
- выяснено, что среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.